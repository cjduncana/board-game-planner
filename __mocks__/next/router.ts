import { NextRouter } from 'next/router'

export const mockPush = jest.fn(() => Promise.resolve(false))

export function useRouter(): NextRouter {
  return {
    route: 'route',
    pathname: 'pathname',
    query: {},
    asPath: 'asPath',
    basePath: 'basePath',
    push: mockPush,
    replace: () => Promise.resolve(false),
    reload: () => undefined,
    back: () => undefined,
    prefetch: Promise.resolve,
    beforePopState: () => undefined,
    events: {
      on: () => undefined,
      off: () => undefined,
      emit: () => undefined,
    },
    isFallback: false,
  }
}
