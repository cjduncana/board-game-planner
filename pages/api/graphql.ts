import { ApolloServer } from 'apollo-server-micro'
import type { PageConfig } from 'next'

import { closeConnectionPlugin, createContext } from '../../lib/context'
import { schema } from '../../lib/schema'

const apolloServer = new ApolloServer({
  context: createContext,
  plugins: [closeConnectionPlugin],
  schema,
})

export const config: PageConfig = { api: { bodyParser: false } }

export default apolloServer.createHandler({ path: '/api/graphql' })
