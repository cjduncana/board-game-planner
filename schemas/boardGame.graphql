query BoardGames($query: String!) {
  boardGames(query: $query) {
    id
    name
    description
    imageUrl
  }
}