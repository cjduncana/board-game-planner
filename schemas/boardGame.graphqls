type BoardGame {
  id: ID!
  name: String!
  description: String!
  minimumAmountOfPlayers: Int!
  maximumAmountOfPlayers: Int!
  thumbnailUrl: String
  imageUrl: String
}