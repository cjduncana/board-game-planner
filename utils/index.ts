import { parse } from 'fast-xml-parser'

export async function fetchXml<Response>(url: string): Promise<Response> {
  const res = await fetch(url)
  const txt = await res.text()
  return parse(txt, { ignoreAttributes: false })
}
