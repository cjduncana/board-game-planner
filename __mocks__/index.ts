import { BaseRouter, CompletePrivateRouteInfo } from 'next/dist/next-server/lib/router/router'
import { NextRouter, Router } from 'next/router'
import React from 'react'

const baseRouter: BaseRouter = {
  route: 'route',
  pathname: 'pathname',
  query: {},
  asPath: 'asPath',
  basePath: 'basePath',
}

export const nextRouter: NextRouter = {
  ...baseRouter,
  push: () => Promise.resolve(false),
  replace: () => Promise.resolve(false),
  reload: () => undefined,
  back: () => undefined,
  prefetch: Promise.resolve,
  beforePopState: () => undefined,
  events: {
    on: (): void => undefined,
    off: (): void => undefined,
    emit: (): void => undefined,
  },
  isFallback: false,
}

const completePrivateRouteInfo: CompletePrivateRouteInfo = {
  Component: React.createElement,
  styleSheets: [],
}

export const router: Router = {
  ...nextRouter,
  components: {},
  sdc: {},
  sub: Promise.resolve,
  clc: null,
  pageLoader: {},
  _bps: undefined,
  _wrapApp: () => undefined,
  isSsr: false,
  onPopState: () => undefined,
  change: () => Promise.resolve(false),
  changeState: () => undefined,
  handleRouteInfoError: () => Promise.resolve(completePrivateRouteInfo),
  getRouteInfo: () => Promise.resolve(completePrivateRouteInfo),
  set: Promise.resolve,
  onlyAHashChange: () => false,
  scrollToHash: () => undefined,
  urlIsNew: () => false,
  _resolveHref: () => ({}),
  fetchComponent: () => Promise.resolve({
    page: React.createElement,
    mod: {},
    styleSheets: [],
  }),
  _getData: Promise.resolve,
  _getStaticData: () => Promise.resolve({}),
  _getServerData: () => Promise.resolve({}),
  getInitialProps: Promise.resolve,
  abortComponentLoad: () => undefined,
  notify: Promise.resolve,
}
