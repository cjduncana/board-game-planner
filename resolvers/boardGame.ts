import { BoardGame, QueryBoardGamesArgs } from '../schemas/type-defs.graphqls'
import { fetchBoardGamesById, searchBoardGameIds } from '../utils/boardGameGeek'

export async function boardGames<Parent>(_parent: Parent, args: QueryBoardGamesArgs): Promise<BoardGame[]> {
  // TODO: Return an empty array if there's no query
  // TODO: Figure out why "7 Wonders: Catan" appears twice
  const gameIds = await searchBoardGameIds(args.query)
  return fetchBoardGamesById(gameIds)
}
