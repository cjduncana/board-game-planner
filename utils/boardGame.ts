import he from 'he'

import { BoardGamesQuery } from '../schemas/boardGame.graphql'

export type BoardGame = BoardGamesQuery['boardGames'][0]

export function getKey(boardGame: BoardGame): string {
  return boardGame.id
}

export function getTitle(boardGame: BoardGame): string {
  return he.decode(boardGame.name)
}

export function getDescription(boardGame: BoardGame): string {
  return he.decode(boardGame.description)
}

export function getImageUrl(boardGame: BoardGame): string | null | undefined {
  return boardGame.imageUrl
}
