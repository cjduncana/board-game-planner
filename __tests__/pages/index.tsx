/**
 * @jest-environment jsdom
 */

import { MockedProvider } from '@apollo/client/testing'
import Grid from '@material-ui/core/Grid'
import Input from '@material-ui/core/Input'
import { RouterContext } from 'next/dist/next-server/lib/router-context'
import { NextRouter } from 'next/router'
import React from 'react'
import TestRenderer from 'react-test-renderer'

import { nextRouter } from '../../__mocks__'
import { baseEvent, createInputEvent } from '../../__mocks__/events'
import Index from '../../pages'
import { mockedBoardGames } from '../../schemas/__mock__/boardGame.graphql'
import { wait } from '../../utils/testHelpers'

// TODO: Fix Image: Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
// TODO: Fix MockedProvider randomly returning the BE result
// Issue is tracked in https://github.com/apollographql/apollo-client/issues/7388
describe('Index page', () => {

  it('should render the page', () => {
    const tree = TestRenderer.create(<Wrapped />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should change the value of the input if the User type a query', () => {
    const { root } = TestRenderer.create(<Wrapped />)

    const input = root.findByType(Input)

    expect(input.props.value).toEqual('')

    TestRenderer.act(() => {
      input.props.onChange(createInputEvent('Catan'))
    })

    expect(input.props.value).toEqual('Catan')
  })

  it('should do nothing if there\'s no new query', () => {
    const { root } = TestRenderer.create(<Wrapped />)

    const form = root.findByType('form')

    expect(mockPush).not.toHaveBeenCalled()

    form.props.onSubmit(baseEvent)

    expect(mockPush).not.toHaveBeenCalled()
  })

  it('should update the router with the new query when the User submit the search', () => {
    const { root } = TestRenderer.create(<Wrapped />)

    const form = root.findByType('form')
    const input = root.findByType(Input)

    TestRenderer.act(() => {
      input.props.onChange(createInputEvent('Catan'))
    })

    expect(mockPush).not.toHaveBeenCalled()

    form.props.onSubmit(baseEvent)

    expect(mockPush).toBeCalledTimes(1)
    expect(mockPush).toHaveBeenCalledWith('pathname?query=Catan', undefined, { shallow: true })
  })

  describe('with a query', () => {

    it('should render the page with results', async () => {
      const component = TestRenderer.create(<Wrapped router={routerWithQuery} />)
      await wait()
      expect(component.toJSON()).toMatchSnapshot()
    })

    it('should search for Board Games if the router has a query', async () => {
      const { root } = TestRenderer.create(<Wrapped router={routerWithQuery} />)

      const gridsBeforeAction = root.findAllByType(Grid)
      expect(gridsBeforeAction).toHaveLength(0)

      await wait()

      const gridsAfterAction = root.findAllByType(Grid)
      expect(gridsAfterAction).toHaveLength(3)
    })

    it('should search for Board Games using the first query if the router has more than one query', async () => {

      const routerWithTwoQueries: NextRouter = {
        ...nextRouter,
        query: { query: ['Catan', 'Pandemic'] },
      }

      const { root } = TestRenderer.create(<Wrapped router={routerWithTwoQueries} />)

      const gridsBeforeAction = root.findAllByType(Grid)
      expect(gridsBeforeAction).toHaveLength(0)

      await wait()

      const gridsAfterAction = root.findAllByType(Grid)
      expect(gridsAfterAction).toHaveLength(3)
    })
  })
})

const mockPush = jest.fn(() => Promise.resolve(false))

const mockedRouter: NextRouter = {
  ...nextRouter,
  push: mockPush,
}

const routerWithQuery: NextRouter = {
  ...nextRouter,
  query: { query: 'Catan' },
}

interface WrappedProps {
  router?: NextRouter
}

function Wrapped(props: WrappedProps): JSX.Element {
  return (
    <RouterContext.Provider value={props.router ?? mockedRouter}>
      <MockedProvider mocks={[mockedBoardGames]}>
        <Index />
      </MockedProvider>
    </RouterContext.Provider>
  )
}
