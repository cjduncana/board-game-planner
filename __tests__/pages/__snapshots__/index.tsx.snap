// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page should render the page 1`] = `
Array [
  <div
    className="makeStyles-heroContent-1"
  >
    <div
      className="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <p
        className="MuiTypography-root MuiTypography-h5 MuiTypography-colorTextSecondary MuiTypography-paragraph MuiTypography-alignCenter"
      >
        Search the available catalog of board games
      </p>
      <form
        onSubmit={[Function]}
      >
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-adornedStart"
          onClick={[Function]}
        >
          <div
            className="MuiInputAdornment-root MuiInputAdornment-positionStart"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <input
            className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
        </div>
      </form>
    </div>
  </div>,
  <span>
    Photo by 
    <a
      href="https://unsplash.com/@jacielmelnik?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText"
      rel="noreferrer"
      target="_blank"
    >
      Jaciel Melnik
    </a>
     on 
    <a
      href="https://unsplash.com/?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText"
      rel="noreferrer"
      target="_blank"
    >
      Unsplash
    </a>
  </span>,
  <div
    style={
      Object {
        "boxSizing": "border-box",
        "display": "block",
        "margin": 0,
        "overflow": "hidden",
        "position": "relative",
      }
    }
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "display": "block",
          "paddingTop": "56.09375%",
        }
      }
    />
    <img
      alt="The Green Meeples"
      decoding="async"
      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
      style={
        Object {
          "border": "none",
          "bottom": 0,
          "boxSizing": "border-box",
          "display": "block",
          "height": 0,
          "left": 0,
          "margin": "auto",
          "maxHeight": "100%",
          "maxWidth": "100%",
          "minHeight": "100%",
          "minWidth": "100%",
          "objectFit": undefined,
          "objectPosition": undefined,
          "padding": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "visibility": "hidden",
          "width": 0,
        }
      }
    />
  </div>,
]
`;

exports[`Index page with a query should render the page with results 1`] = `
Array [
  <div
    className="makeStyles-heroContent-1"
  >
    <div
      className="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <p
        className="MuiTypography-root MuiTypography-h5 MuiTypography-colorTextSecondary MuiTypography-paragraph MuiTypography-alignCenter"
      >
        Search the available catalog of board games
      </p>
      <form
        onSubmit={[Function]}
      >
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-adornedStart"
          onClick={[Function]}
        >
          <div
            className="MuiInputAdornment-root MuiInputAdornment-positionStart"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <input
            className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
        </div>
      </form>
    </div>
  </div>,
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
    >
      <div
        className="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          className="MuiCardMedia-root makeStyles-cardMedia-3"
          style={
            Object {
              "backgroundImage": "url(\\"https://cf.geekdo-images.com/W3Bsga_uLP9kO91gZ7H8yw__original/img/A-0yDJkve0avEicYQ4HoNO-HkK8=/0x0/pic2419375.jpg\\")",
            }
          }
          title="Catan"
        />
        <div
          className="MuiCardContent-root makeStyles-cardContent-4"
        >
          <h2
            className="MuiTypography-root makeStyles-title-5 MuiTypography-h5 MuiTypography-gutterBottom"
          >
            Catan
          </h2>
          <p
            className="MuiTypography-root makeStyles-description-6 MuiTypography-body1"
          >
            In CATAN (formerly The Settlers of Catan), players try to be the dominant force on the island of Catan by building settlements, cities, and roads. On each turn dice are rolled to determine what resources the island produces. Players build by spending resources (sheep, wheat, wood, brick and ore) that are depicted by these resource cards; each land type, with the exception of the unproductive desert, produces a specific resource: hills produce brick, forests produce wood, mountains produce ore, fields produce wheat, and pastures produce sheep.&#10;&#10;Setup includes randomly placing large hexagonal tiles (each showing a resource or the desert) in a honeycomb shape and surrounding them with water tiles, some of which contain ports of exchange. Number disks, which will correspond to die rolls (two 6-sided dice are used), are placed on each resource tile. Each player is given two settlements (think: houses) and roads (sticks) which are, in turn, placed on intersections and borders of the resource tiles. Players collect a hand of resource cards based on which hex tiles their last-placed house is adjacent to. A robber pawn is placed on the desert tile.&#10;&#10;A turn consists of possibly playing a development card, rolling the dice, everyone (perhaps) collecting resource cards based on the roll and position of houses (or upgraded cities&mdash;think: hotels) unless a 7 is rolled, turning in resource cards (if possible and desired) for improvements, trading cards at a port, and trading resource cards with other players. If a 7 is rolled, the active player moves the robber to a new hex tile and steals resource cards from other players who have built structures adjacent to that tile.&#10;&#10;Points are accumulated by building settlements and cities, having the longest road and the largest army (from some of the development cards), and gathering certain development cards that simply award victory points. When a player has gathered 10 points (some of which may be held in secret), he announces his total and claims the win.&#10;&#10;CATAN has won multiple awards and is one of the most popular games in recent history due to its amazing ability to appeal to experienced gamers as well as those new to the hobby.&#10;&#10;Die Siedler von Catan was originally published by KOSMOS and has gone through multiple editions. It was licensed by Mayfair and has undergone four editions as The Settlers of Catan. In 2015, it was formally renamed CATAN to better represent itself as the core and base game of the CATAN series. It has been re-published in two travel editions, portable edition and compact edition, as a special gallery edition (replaced in 2009 with a family edition), as an anniversary wooden edition, as a deluxe 3D collector's edition, in the basic Simply Catan, as a beginner version, and with an entirely new theme in Japan and Asia as Settlers of Catan: Rockman Edition. Numerous spin-offs and expansions have also been made for the game.&#10;&#10;
          </p>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
    >
      <div
        className="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <span
          className="MuiSkeleton-root MuiSkeleton-text makeStyles-cardMedia-3 MuiSkeleton-pulse"
          style={
            Object {
              "height": undefined,
              "width": undefined,
            }
          }
        />
        <div
          className="MuiCardContent-root makeStyles-cardContent-4"
        >
          <h2
            className="MuiTypography-root makeStyles-title-5 MuiTypography-h5 MuiTypography-gutterBottom"
          >
            Catan: First Adventure Print & Play
          </h2>
          <p
            className="MuiTypography-root makeStyles-description-6 MuiTypography-body1"
          >
            The kids on Catan are very adventurous. They learn to sail at a very young age and like to spend their time sailing around the small offshore islands and playing exciting games. Today you are playing pirates.&#10;&#10;You and your friends race to build small pirate lairs and, ultimately, a great pirate fort. The first to do so will be crowned the ruler of the pirates!&#10;&#10;For this game, one friend pretends to be a Ghost Captain and tries to prevent the others from completing their lairs. Who among you will be the most powerful pirate?&#10;&#10;Print, craft, and play this special kids' game, which is released as a part of the CATAN #StayAndPlay campaign, which has the goal of making home-staying during the COVID-19 pandemic a little more fun.&#10;&#10;
          </p>
        </div>
      </div>
    </div>
  </div>,
]
`;
